### ISSUE TITLE
CORS issues on preload nodes

### ISSUE THREAD

**Issue by daviddias** [ipfs/infrastructure#447]
Hi Infra team, some of our users are reporting having multiple problems with CORS, can you confirm if this is a known problem?

Issue: https://github.com/ipfs/js-ipfs/issues/1476#issuecomment-434741386

**Comment by alanshaw** [comment]
This is still happening and can be see here https://klueq.github.io/ (inspect the browser console)

**Comment by alanshaw** [comment]
https://github.com/ipfs/js-ipfs/issues/1694

**Comment by haadcode** [comment]
Adding my two cents to double confirm we're seeing this issue also, described in https://github.com/ipfs/js-ipfs/issues/1694.

**Comment by diasdavid** [comment]
Another report with example - https://discuss.ipfs.io/t/trouble-with-content-availability-on-public-gateways-from-js-ipfs/4137/1

I tested with Chrome and Firefox. Trimmed the test case to: https://codepen.io/anon/pen/jQqMPb?editors=0011

**Comment by lidel** [comment]
### Potential Source of Issues

Preload nodes are sometimes slow to respond which translates to a persistent problem of [preflight requests (HTTP OPTIONS)](https://developer.mozilla.org/en-US/docs/Glossary/Preflight_request) failing, that failure being cached by the browser and as a result every following preload  request (HTTP GET to `/api/v0/refs`) is blocked by the browser:

>  ![2018-11-08--15-59-45](https://user-images.githubusercontent.com/157609/48207147-5499ad80-e370-11e8-8b8c-342dadf12bed.png)


### Potential Fix

@kyledrake what if we introduce an easy optimization: detect and return static response for HTTP OPTIONS requests at Nginx.

That way preflight requests will not fail due to go-ipfs being under load, because they will never hit the daemon.

Static response to copy:

```console
$ curl -X OPTIONS 'https://node0.preload.ipfs.io/api/v0/refs' -v
# (...)
< HTTP/1.1 200 OK
< Date: Thu, 08 Nov 2018 15:09:18 GMT
< Content-Length: 0
< Connection: keep-alive
< Vary: Origin
< Vary: Access-Control-Request-Method
< Vary: Access-Control-Request-Headers
< Strict-Transport-Security: max-age=31536000; includeSubDomains; preload
< Access-Control-Allow-Origin: *
< Access-Control-Allow-Methods: GET, POST, OPTIONS
< Access-Control-Allow-Headers: X-Requested-With, Range, Content-Range, X-Chunked-Output, X-Stream-Output
< Access-Control-Expose-Headers: Content-Range, X-Chunked-Output, X-Stream-Output
```

**Comment by daviddias** [comment]
@ipfs/infrastructure any update here?

**Comment by kyledrake** [comment]
@daviddias

I deployed new preload servers that put them on their own dedicated machines and brings them up to the latest version of IPFS, which improved things a lot. We also had someone flooding the preload server with several hundred requests per second, which was causing performance issues. I throttled it so that people can't just flood the server with API calls from a script.

I'm going to explore @lidel's idea here when I get a chance, but I'd like to keep this open until I deploy that.

**Comment by mitra42** [comment]
If you've throttled it - totally reasonable - I hope that the JS-IPFS code has been written so that if preload is slow responding (throttled) that it won't wait for it before trying the local JS-IPFS (not JS-IPFS-API) call.  

Note that it might not be a script thats doing a fast load, if we load a webpage then its going to try and read up to approx 50 images. In many cases those will be cached but sometimes the js-ipfs is going to be inserting these preload calls. 

Note - its on our list to turn preload off in the dweb.archive.org code, but we haven't got to it yet.

**Comment by lidel** [comment]
@mitra42  afaik preload in js-ipfs is async and does not block regular operations:

> Preload API requests are now done asynchronously so they don't effect the time it takes to add content. The preload module keeps track of in flight requests and cancels any that are still flying when stop is called on the node. – https://github.com/ipfs/js-ipfs/pull/1464#issuecomment-408111913

@kyledrake setting an explicit timeout for caching preflight requests could limit the number of preflight requests for popular CIDs (preflight cache is per URL) and ensure all vendors behave the same, eg. to set maximum possible cache of 10 minutes:

```console
Access-Control-Max-Age: 600
```
 
> <dl>
> <dt>Access-Control-Max-Age</dt>
> 
> <dd>Maximum number of seconds the results can be cached.
>  Firefox caps this at 24 hours (86400 seconds) and Chromium at 10 minutes (600 seconds). Chromium also specifies a default value of 5 seconds.
> A value of -1 will disable caching, requiring a preflight OPTIONS check for all calls.</dd>
> </dl>



Refs:
- https://www.w3.org/TR/cors/#preflight-result-cache-0
- https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Access-Control-Max-Age

**Comment by mitra42** [comment]
Just confirming, I'm still seeing preload CORS errors https://github.com/ipfs/infrastructure/issues/447

**Comment by eefahy** [comment]
paging @kyledrake

**Comment by alanshaw** [comment]
Reported here also https://github.com/ipfs/js-ipfs/issues/1732

**Comment by kyledrake** [comment]
There was a known issue with the AMS server which I just resolved, but the other servers should be operating as expected. The monitoring caught the AMS outage, but I didn't see issues with any of other servers.

Having all of the bootstrap servers go down at once is of course a major issue, but if it's one out of several, that shouldn't cause any major problems (sans the annoying console output). We're going to have random servers go down from time to time for various reasons.

I'm going to close this ticket, as the preload servers CORS is functioning properly. I am working on improving the infrastructure behind the current bootstrap system to make it faster and more reliable in the future, but that's work that's independent of this ticket (which is specifically about CORS issues).

**Comment by mitra42** [comment]
Not sure what you mean by API endpoint, I see this in a browser log, and as mentioned by other reporters its intermittent - and I'm usually only looking at the browser log if Im looking for some other problem. If there is other info I can grab from you next time I see it in a log please let me know (here).

**Comment by lidel** [comment]
@mitra42 (endpoint == server with `/api/v0/`) I am mainly curious if CORS errors in console look like [this](https://user-images.githubusercontent.com/157609/48207147-5499ad80-e370-11e8-8b8c-342dadf12bed.png) or something else. Next time you see something just grab a screenshot and post here or in a new issue and @lidel me.

**Comment by mitra42** [comment]
Ok - will do - do you mean something different from that posted in the bug report on [js-ipfs#1476](https://github.com/ipfs/js-ipfs/issues/1476). There is a bunch of history to this problem there which of course didn't get copied over when this new one was created.
